### PR TITLE
Update control-plane-flags.md

### PR DESCRIPTION
**Pull request by kigawas**
Add details about updating configuration via kubeadm

### CODE REVIEW

File: content/en/docs/setup/independent/control-plane-flags.md [hunk 1]
```diff
@@ -21,6 +21,9 @@ The `extraArgs` field consist of `key: value` pairs. To override a flag for a co
 
 1.  Add the appropriate fields to your configuration.
 2.  Add the flags to override to the field.
+3.  Update the configuration by `kubeadm upgrade apply --config config.yml`
```
Comment (sftim) [inline content/en/docs/setup/independent/control-plane-flags.md]:
```suggestion
3.  Update the configuration by running `kubeadm upgrade apply --config config.yml` 
```
What do you think @kigawas?

File: content/en/docs/setup/independent/control-plane-flags.md [hunk 2]
```diff
@@ -21,6 +21,9 @@ The `extraArgs` field consist of `key: value` pairs. To override a flag for a co
 
 1.  Add the appropriate fields to your configuration.
 2.  Add the flags to override to the field.
+3.  Update the configuration by `kubeadm upgrade apply --config config.yml` 
+
+You can also add `extraVolumes` to mount additional volumes if necessary (e.g. specifying a configuration file for cloud provider gce).
```
Comment (sftim) [inline content/en/docs/setup/independent/control-plane-flags.md]:
```suggestion
You can also add `extraVolumes` to mount additional volumes if necessary (e.g. including configuration files that are relevant to your cloud provider).
```

File: content/en/docs/setup/independent/control-plane-flags.md [hunk 3]
```diff
@@ -64,6 +67,13 @@ controllerManager:
     cluster-signing-key-file: /home/johndoe/keys/ca.key
     bind-address: 0.0.0.0
     deployment-controller-sync-period: 50
+    cloud-provider: gce
```
Comment (sftim) [inline content/en/docs/setup/independent/control-plane-flags.md]: These docs aim to be vendor-neutral, so how about
```suggestion
    cloud-provider: exampleprovider
```
?

I wonder if it makes sense to have two tabs, one that has the _extraVolumes_ and the cloud provider set and another that doesn't? I know that cluster operators definitely deploy with `kubeadm` onto bare metal, and that that's supported.

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 1 above)
Comment (neolit123) [reply content/en/docs/setup/independent/control-plane-flags.md]: -1 on including `kubeadm upgrade apply --config config.yml`.
applying upgrade for the sake of modifying a flag is possible, but not recommended.

there are plans to support simpler customizations eventually.

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 3 above)
Comment (neolit123) [reply content/en/docs/setup/independent/control-plane-flags.md]: yeah, cloud provider examples should not be included in this document.
see https://kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#kubeadm

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 2 above)
Comment (neolit123) [reply content/en/docs/setup/independent/control-plane-flags.md]: there are already examples here:
https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2#hdr-Kubeadm_init_configuration_types

the godoc ref pages are linked in this document already.

**Comment by neolit123** [comment]
@kigawas 
thanks for the PR, but i don't think we should include some of these changes. there is already overlap with the godoc API ref. docs.

/hold

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 1 above)
Comment (kigawas) [reply content/en/docs/setup/independent/control-plane-flags.md]: > -1 on including `kubeadm upgrade apply --config config.yml`.
> applying upgrade for the sake of modifying a flag is possible, but not recommended.
> 
> there are plans to support simpler customizations eventually.

Thanks for pointing out.

Can you make a recommendation of modifying flags at the current 14.2 version?

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 2 above)
Comment (kigawas) [reply content/en/docs/setup/independent/control-plane-flags.md]: > there are already examples here:
> https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2#hdr-Kubeadm_init_configuration_types
> 
> the godoc ref pages are linked in this document already.

This link is good, but not in this document.

The link in this document is https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm#ClusterConfiguration

which is less informative

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 1 above)
Comment (neolit123) [reply content/en/docs/setup/independent/control-plane-flags.md]: with 1.14 and 1.15 you must modify the `/etc/kubernetes/manifests/` YAML files manually, but these changes will be lost after upgrade. to make the changes persist you must also modify the `kubeadm-config` ConfigMap object.

`kubeadm upgrade apply --config config.yml` solves this problem but it also performs all the steps of an upgrade which might not be desired by some users.

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 2 above)
Comment (neolit123) [reply content/en/docs/setup/independent/control-plane-flags.md]: i think it's fine to change the link to https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 1 above)
Comment (kigawas) [reply content/en/docs/setup/independent/control-plane-flags.md]: Thanks, so a user may need to modify files under `/etc/kubernetes/manifests/` and both `kubeadm-config` configmap?

I've done some search, it looks like this point is seldom mentioned. Shall we document it here?

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 3 above)
Comment (kigawas) [reply content/en/docs/setup/independent/control-plane-flags.md]: Agree not to include cloud provider here

Comment (kigawas) [reply content/en/docs/setup/independent/control-plane-flags.md]: @neolit123 

By the way, I noticed that `--cloud-provider` is already deprecated when starting kubelet and kube-controller-manager, and it said the in-tree cloud providers will be moved to out-tree repos like https://github.com/kubernetes/cloud-provider-gcp

If I want to integrate the out-tree cloud provider code above, do you know how I can do this?

Comment (neolit123) [reply content/en/docs/setup/independent/control-plane-flags.md]: unfortunately no. if you have GCP specific question you can try asking in the #sig-gcp slack channel.

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 1 above)
Comment (neolit123) [reply content/en/docs/setup/independent/control-plane-flags.md]: no, this should not be documented here.

File: content/en/docs/setup/independent/control-plane-flags.md [hunk 4]
```diff
@@ -21,9 +21,14 @@ The `extraArgs` field consist of `key: value` pairs. To override a flag for a co
...
 2.  Add the flags to override to the field.
+3.  Update the configuration by
+    1. Editing configuration files under `/etc/kubernetes/manifests/` directory
+    2. Editing the kubeadm configmap by running `kubectl -n kube-system edit configmaps kubeadm-config`
+
+You can also add `extraVolumes` to mount additional volumes if necessary (e.g. including configuration files that are relevant to your cloud provider).
 
 For more details on each field in the configuration you can navigate to our
-[API reference pages](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm#ClusterConfiguration).
+[API reference pages](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2#hdr-Kubeadm_init_configuration_types).
```
Comment (neolit123) [inline content/en/docs/setup/independent/control-plane-flags.md]: this should be:
> [API reference pages](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta2)

i think this is the only change this PR should do for now.

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 3 above)
Comment (kigawas) [reply content/en/docs/setup/independent/control-plane-flags.md]: > unfortunately no. if you have GCP specific question you can try asking in the #sig-gcp slack channel.

I asked several days ago, but the channel is not active and there is no one replying me

File: content/en/docs/setup/independent/control-plane-flags.md [hunk 5]
```diff
@@ -22,8 +22,10 @@ The `extraArgs` field consist of `key: value` pairs. To override a flag for a co
 1.  Add the appropriate fields to your configuration.
 2.  Add the flags to override to the field.
 
+You can also add `extraVolumes` to mount additional volumes if necessary (e.g. including configuration files that are relevant to your cloud provider).
+
```
Comment (neolit123) [inline content/en/docs/setup/independent/control-plane-flags.md]: please exclude this sentence for now.
we might refactor this document eventually.

File: content/en/docs/setup/independent/control-plane-flags.md [hunk 6]
```diff
@@ -22,8 +22,10 @@ The `extraArgs` field consist of `key: value` pairs. To override a flag for a co
...
 For more details on each field in the configuration you can navigate to our
-[API reference pages](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm#ClusterConfiguration).
+[API reference pages](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1#hdr-Kubeadm_init_configuration_types).
```
Comment (neolit123) [inline content/en/docs/setup/independent/control-plane-flags.md]: should be only `[API reference pages](https://godoc.org/k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm/v1beta1)`

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 3 above)
Comment (neolit123) [reply content/en/docs/setup/independent/control-plane-flags.md]: you can also try the #sig-cloud-provider channel.

File: content/en/docs/setup/independent/control-plane-flags.md (hunk 5 above)
Comment (kigawas) [reply content/en/docs/setup/independent/control-plane-flags.md]: Okay it's done

**Comment by tengqm** [comment]
How can I merge the discussions as well? They are so inspiring. :)
/approve

**Pull request merged**